File: queue/queue.cpp
#include <iostream>

struct Node
{
    Node * PrevNode = nullptr;
    Node * NextNode = nullptr;
    int Data;
    Node(int InData = 0)
    {
        PrevNode = nullptr;
        NextNode = nullptr;
        Data = InData;
    }
};

void SafeDelete(Node * InNode)
{
    if(InNode)
    {
        delete InNode;
        InNode = nullptr;
    }
}

class Queue
{
private:
    Node * HeadNode;
    Node * TailNode;
    int ElementCount;

public:
    Queue();
    ~Queue();
    bool IsEmpty();
    bool PushData(int InData);
    int PopData();
    int GetSize();
    void PrintData();
};

Queue::Queue()
{
    HeadNode = nullptr;
    TailNode = nullptr;
    ElementCount = 0;
}

Queue::~Queue()
{
    if(TailNode)
    {
        Node * TempNode = TailNode->PrevNode;
        while(TempNode)
        {
            Node * DeleteNode = TempNode;
            TempNode = TempNode->PrevNode;
            SafeDelete(DeleteNode);
        }

        TailNode = nullptr;
    }

    HeadNode = nullptr;
}

bool Queue::IsEmpty()
{
    return ElementCount == 0;
}

bool Queue::PushData(int InData)
{
    if(HeadNode == nullptr)
    {
        Node * PushNode = new Node(InData);
        HeadNode = PushNode;
        TailNode = HeadNode;
    }
    else
    {
        Node * PushNode = new Node(InData);
        Node * TempNode = this->HeadNode;
        TempNode->PrevNode = PushNode;
        PushNode->NextNode = TempNode;
        this->HeadNode = PushNode;
    }

    ElementCount++;

    return true;
}

int Queue::PopData()
{
    if(ElementCount == 0)
    {
        std::cout << "큐에 데이터가 없습니다." << std::endl;
        return -1;
    }

    int Data = 0;
    Node * PopNode = HeadNode;
    Data = PopNode->Data;
    PopNode->NextNode->PrevNode = nullptr;
    HeadNode = PopNode->NextNode;
    SafeDelete(PopNode);
    ElementCount--;

    return Data;
}

int Queue::GetSize()
{
    return ElementCount;
}

void Queue::PrintData()
{
    if(ElementCount == 0)
    {
        std::cout << "출력할 데이터가 없습니다." << std::endl;
    }

    int Count = 0;
    Node * TempNode = this->HeadNode;
    while(TempNode)
    {
        std::cout << Count + 1 << "번째 데이터 : " << TempNode->Data << std::endl; 
        TempNode = TempNode->NextNode;
        Count++;
    }
}

int main(){
    
    Queue QueueList;

    std::cout << "큐의 현재 사이즈 :" << QueueList.GetSize() << std::endl; 

    for(int i=0; i<11; ++i)
    {
        QueueList.PushData(i);
    }

    QueueList.PrintData();
    std::cout << "큐의 현재 사이즈 :" << QueueList.GetSize() << std::endl; 
    
    for(int j=0; j<5; ++j)
    {
        std::cout << "데이터 : " << QueueList.PopData() << " 팝업" << std::endl;
    }

    QueueList.PrintData();
    std::cout << "큐의 현재 사이즈 :" << QueueList.GetSize() << std::endl; 
}

File: queue/queue.py
#! /usr/bin/python
# -*- coding:utf-8 -*-


class Node:
    def __init__(self, data):
        self.data = data
        self.prevNode = None
        self.nextNode = None


class Queue:
    def __init__(self):
        self.headNode = None
        self.tailNode = None
        self.elementCount = 0

    def GetSize(self):
        return self.elementCount

    def PrintData(self):
        if self.headNode is None:
            print("출력할 데이터가 없습니다.")
            return
        count = 0
        currentNode = self.headNode
        while currentNode:
            count += 1
            print(count, '번째 데이터 : ', currentNode.data)
            currentNode = currentNode.nextNode
            if currentNode is self.headNode:
                return

    def Enqueue(self, data):
        insertNode = Node(data)
        if self.headNode is None:
            self.headNode = insertNode
            self.tailNode = self.headNode
        else:
            currentNode = self.headNode
            currentNode.prevNode = insertNode
            insertNode.nextNode = currentNode
            self.headNode = insertNode

        self.headNode.prevNode = self.tailNode
        self.tailNode.nextNode = self.headNode
        self.elementCount += 1

    def Dequeue(self):
        if self.headNode is None:
            print("데이터가 없습니다.")

        popNode = self.tailNode
        data = popNode.data
        if popNode == self.headNode:
            self.HeadNode = None
            self.TailNode = None
        else:
            self.tailNode = popNode.prevNode
            self.tailNode.nextNode = self.headNode
            self.headNode.prevNode = self.tailNode
        popNode = None
        self.elementCount -= 1
        return data


if __name__ == '__main__':
    queue = Queue()
    queue.PrintData()
    print('큐 사이즈 : ', queue.GetSize())

    queue.Enqueue(10)
    queue.PrintData()
    print('큐 사이즈 : ', queue.GetSize())

    for i in range(1, 10):
        queue.Enqueue(i)
    queue.PrintData()
    print('큐 사이즈 : ', queue.GetSize())

    print('Dequeue : ', queue.Dequeue())
    print('Dequeue : ', queue.Dequeue())
    queue.PrintData()
    print('큐 사이즈 : ', queue.GetSize())


